neco-server-exporter/pkg/exporter/exporter.go

```diff
@@ -0,0 +1,102 @@
+package exporter
+
+import (
+	"context"
+	"errors"
+	"log/slog"
+	"net/http"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/VictoriaMetrics/metrics"
+)
+
+type Exporter struct {
+	running    atomic.Bool
+	interval   time.Duration
+	collectors []Collector
+
+	log *slog.Logger
+}
+
+func NewExporter(interval time.Duration) *Exporter {
+	return &Exporter{
+		interval:   interval,
+		collectors: make([]Collector, 0),
+		log:        slog.New(slog.NewJSONHandler(os.Stdout, nil)),
+	}
+}
+
+func (e *Exporter) AddCollector(c Collector) error {
+	if e.running.Load() {
+		return errors.New("shouuld add collector before start")
+	}
+	e.collectors = append(e.collectors, c)
+	return nil
+}
+
+func (e *Exporter) Start(ctx context.Context) error {
+	go func() {
+		if err := e.Run(ctx); err != nil {
+			e.log.Error("failed to scrape metrics", slog.Any("error", err))
+		}
+	}()
+
+	http.HandleFunc("/metrics", func(w http.ResponseWriter, req *http.Request) {
+		metrics.WritePrometheus(w, false)
+	})
+	e.log.InfoContext(ctx, "start metrics server")
+
+	if err := http.ListenAndServe("0.0.0.0:8080", nil); err != nil {
+		e.log.Error("metrics server stopped", slog.Any("error", err))
+		return err
+	}
+	return nil
+}
+
+func (e *Exporter) Run(ctx context.Context) error {
```

Comment:
It seems that `Run()` never return an error.

Reply:
I think it's better to leave it as is, because it states the Run function may return error in future development.
Without it, someone may confuse how to treat errors happen in it.

Reply:
That makes sense. How about writing it in a comment to left what you replied?
For example...

```go
// Run starts the metric collection loop.
// The error return is reserved for future fatal error conditions.
// Currently always returns nil except when context is cancelled.
func (e *Exporter) Run(ctx context.Context) error {
```